log: net: Add TCP support to syslog transmitter

jukkar opened the pull request:
Allow user to configure the syslog net backend to use TCP
instead of UDP. The syslog server address for TCP needs to
have "tcp://" URI in front of the address, for example the
tcp://192.0.2.2 server address would use TCP as a transport.
If there is no URI, then UDP would be used by default.
    
See the relevant RFC at https://www.rfc-editor.org/rfc/rfc6587
for details.
    
Fixes #66728

jukkar committed:
log: net: Add TCP support to syslog transmitter

Allow user to configure the syslog net backend to use TCP
instead of UDP. The syslog server address for TCP needs to
have "tcp://" URI in front of the address, for example the
tcp://192.0.2.2 server address would use TCP as a transport.
If there is no URI, then UDP would be used by default.

See the relevant RFC at https://www.rfc-editor.org/rfc/rfc6587
for details.

Fixes #66728

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
log: net: Depend on full featured libc instead of just newlibc

For time printout, depend on full featured libc like picolibc
instead of just newlibc.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -92,7 +111,12 @@ static int do_net_init(struct log_backend_net_ctx *ctx)
 	local_addr->sa_family = server_addr.sa_family;
 
-	ret = zsock_socket(server_addr.sa_family, SOCK_DGRAM, IPPROTO_UDP);
+	if (ctx->is_tcp) {
```
Maybe some explicit error log about misconfiguration could be printed in case  `CONFIG_NET_TCP` is disabled?

rlubos commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -48,12 +46,33 @@ static int line_out(uint8_t *data, size_t length, void *output_ctx)
+#if defined(CONFIG_NET_TCP)
+	char len[sizeof("123456789")];
+
+	if (ctx->is_tcp) {
```
Could this be a runtime check, ie `IS_ENABLED(CONFIG_NET_TCP) && ctx->is_tcp`?

rlubos replied:
Was this overlooked in the recent push?

jukkar replied:
There is the `CONFIG_NET_TCP` few lines above but I can certainly put those inside `IS_ENABLED`, dunno why I did it like that.

jukkar replied:
Remembered now why it was done without IS_ENABLED, the len var is allocated from stack so it cannot be inside a block

rlubos replied:
Hmm, I just wonder if it wouldn't be compiled out if not used, even if declared outside of the block. But in the end, it's nothing important so we can leave as is as well.

jukkar replied:
If using IS_ENABLED() we would referece invalid pointer in sendmsg() call

rlubos commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -48,12 +46,33 @@ static int line_out(uint8_t *data, size_t length, void *output_ctx)
+	msg.msg_iov = io_vector;
+	msg.msg_iovlen = pos;
+
+	ret = zsock_sendmsg(ctx->sock, &msg, ZSOCK_MSG_DONTWAIT);
```
Does it sill make sense to use non-blocking send with TCP? It's no longer best-effort as with UDP, we should keep the stream consistent. And with TCP it's easier to get EAGAIN as a result (by filling TX window).

jukkar replied:
Not sure if we should block here. I mean the logging would get delayed in that case which could be problematic if synchronous logging is used. And if we cannot send, we should drop the message anyway.

rlubos replied:
And wouldn't it be a problem that we send the message length first, and then incomplete message (sorry, I didn't dig into RFC yet). With TCP it's not all or nothing, we could just send part of the buffer.

jukkar replied:
I was trying to avoid few byte send that was happening if I had two sends.

jukkar committed:
log: net: Add structured logging support to syslog

Add structured logging support to syslog backend. This means
that meta data can be included to logging output.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* I added structured data support
* Removed `MSG_DONTWAIT` for TCP socket

jukkar commented:
* Noticed an issue in the structured data values, they must be always inside `""` according to RFC.

jukkar commented:
> But timeQuality cannot return constant values. There must be either a callback or some variable in which the rest of the applications/drivers will set the status of whether the RTC is currently properly synchronized.
> Returning a constant value from config is not good.

Yes, using constant values here seemed wrong to me, problem with timeQuality settings is that from where to get the values for these. I am not sure if we have suitable time APIs to dig these values.

psztoch commented:
isSynced flag is very important for any embedded system.
Of course its implementation may depend on many factors and it will be implemented differently in different applications.

Nevertheless, every system (including zephyr) should know whether it works based on a certain time (isSynced=true)...

Systems that do not have RTC may set this flag after the first synchronization with the time server.
Systems that have RTCs can return TRUE when everything with the RTC is OK and it has been initialized correctly or when the last synchronization of this RTC with the time server was no more than, for example, a week ago.

Nevertheless, I believe that such a standardized flag in the system is needed. When isSynced=false, some functionalities (e.g. related to the calendar) should be automatically disabled.

jukkar commented:
Trying to kick this PR forward, @pdgendt would you have some time to check it?

pdgendt commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -48,12 +46,33 @@ static int line_out(uint8_t *data, size_t length, void *output_ctx)
+		io_vector[pos].iov_len = strlen(len);
+		pos++;
+	}
+#endif
```
Do we need an else case?

```c
#if defined(CONFIG_NET_TCP)
// ...
#else
	if (ctx->is_tcp) {
		return -ENOTSUP;
	}
#endif
```

jukkar replied:
True, it would make sense here. Will fix.

pdgendt commented:
A nice addition would be to allow to set this at runtime, but this is out of the scope of this PR.

jukkar commented:
> A nice addition would be to allow to set this at runtime, but this is out of the scope of this PR.

Yep, support for runtime configuration could be added in future PRs.

jukkar commented:
ping @nordic-krch, PTAL
